timebp/nanosecond.go

```diff
@@ -0,0 +1,116 @@
+package timebp
+
+import (
+	"encoding/json"
+	"strconv"
+	"time"
+)
+
+var (
+	_ json.Marshaler   = TimestampNanosecond{}
+	_ json.Unmarshaler = (*TimestampNanosecond)(nil)
+	_ json.Marshaler   = DurationNanosecond(0)
+	_ json.Unmarshaler = (*DurationNanosecond)(nil)
+)
+
+const nanosecondsPerSecond = int64(time.Second / time.Nanosecond)
+
+// TimestampNanosecond implements json encoding/decoding using nanoseconds
+// since EPOCH.
+type TimestampNanosecond time.Time
+
+func (ts TimestampNanosecond) String() string {
+	return ts.ToTime().String()
+}
+
+// ToTime converts TimestampNanosecond back to time.Time.
+func (ts TimestampNanosecond) ToTime() time.Time {
+	return time.Time(ts)
+}
+
+// MarshalJSON implements json.Marshaler interface, using epoch nanoseconds.
+func (ts TimestampNanosecond) MarshalJSON() ([]byte, error) {
+	t := ts.ToTime()
+	if t.IsZero() {
+		return []byte("null"), nil
+	}
+
+	return []byte(strconv.FormatInt(TimeToNanoseconds(t), 10)), nil
+}
+
+// UnmarshalJSON implements json.Unmarshaler interface.
+func (ts *TimestampNanosecond) UnmarshalJSON(data []byte) error {
+	s := string(data)
+	if s == "null" {
+		*ts = TimestampNanosecond{}
+		return nil
+	}
+
+	us, err := strconv.ParseInt(s, 10, 64)
+	if err != nil {
+		return err
+	}
+	*ts = TimestampNanosecond(NanosecondsToTime(us))
+	return nil
+}
+
+// NanosecondsToTime converts milliseconds since EPOCH to time.Time.
```

Comment:
I suppose this should be `nanoseconds`

Reply:
yeah, i missed renamed the variable "us" to "ns"

Reply:
@fishy thanks for aclaration, please see my response below